### PR TITLE
feat: add Nx.Random.shuffle

### PR DESCRIPTION
**Pull request by polvalente**
Based on jax.random.permutation, so I can implement jax.random.choice aftewards

### CODE REVIEW

**Comment by jonatanklosko** [comment]
@polvalente we do have `Nx.shuffle`, so we need to standardize somehow.

**Comment by polvalente** [comment]
@jonatanklosko I'm ok with exposing shuffle here instead of permutation. I just followed the name that jax uses

File: nx/lib/nx/random.ex [hunk 1]
```diff
@@ -526,6 +526,80 @@ defmodule Nx.Random do
...
 
+  @doc """
+  Randomly shuffles a tensor along an axis
+
+  ## Options
+
+    * `:axis` - the axis along which to shuffle. Defaults to `0`
+    * `:independent` - a boolean that indicates wether the permutations
+      are independent along the given axis. Defaults to `false`
+
```
Comment (jonatanklosko) [inline nx/lib/nx/random.ex:537]:
```suggestion
    * `:axis` - the axis along which to shuffle. Defaults to `0`

    * `:independent` - a boolean that indicates wether the permutations
      are independent along the given axis. Defaults to `false`

```

**Comment by jonatanklosko** [comment]
Yeah, I meant more that we should deprecate `Nx.shuffle`. Also note that id behaves differently when no axis is given.

**Comment by josevalim** [comment]
Good call. We _must_ deprecate shuffle, because it uses random_uniform. I will add a deprecation warning shortly.

**Comment by josevalim** [comment]
Actually, I can't deprecate it because we need this PR :P @polvalente can you add

```
@doc type: :creation, deprecated: "Use Nx.Random.shuffle/2 instead"
```

to the top of shuffle pls?

File: nx/lib/nx/random.ex [hunk 2]
```diff
@@ -526,6 +526,81 @@ defmodule Nx.Random do
     end)
   end
 
+  @doc """
+  Randomly shuffles a tensor along an axis
```
Comment (jonatanklosko) [inline nx/lib/nx/random.ex:530]:
```suggestion
  Randomly shuffles a tensor elements along an axis.
```

File: nx/lib/nx/random.ex [hunk 3]
```diff
@@ -526,6 +526,81 @@ defmodule Nx.Random do
...
+      >
+  """
+  defn shuffle(key, tensor, opts \\ []) do
+    opts = keyword!(opts, axis: 0, independent: false)
+    axis = opts[:axis]
+
+    if opts[:independent] do
+      shuffle_independent(key, tensor, axis: axis)
+    else
+      {idx, key} = shuffle_independent(key, Nx.iota({Nx.axis_size(tensor, axis)}), axis: axis)
```
Comment (jonatanklosko) [inline nx/lib/nx/random.ex:572]:
```suggestion
      {idx, key} = shuffle_independent(key, Nx.iota({Nx.axis_size(tensor, axis)}), axis: 0)
```

Comment (polvalente) [reply nx/lib/nx/random.ex:572]: gotta add a test for that

**Review by jonatanklosko — APPROVED** [review]
